External/FEXCore/Source/Interface/Core/JIT/Arm64/JIT.cpp

```diff
@@ -813,64 +804,90 @@ void *JITCore::CompileCode([[maybe_unused]] FEXCore::IR::IRListView<true> const
   auto Buffer = GetBuffer();
   auto Entry = Buffer->GetOffsetAddress<uint64_t>(GetCursorOffset());
 
+ if (CTX->GetGdbServerStatus()) {
+    aarch64::Label RunBlock;
 
-  if (SpillSlots) {
-    add(TMP1, sp, 0); // Move that supports SP
-    sub(sp, sp, SpillSlots * 16);
-    stp(TMP1, lr, MemOperand(sp, -16, PreIndex));
-  }
-  else {
-    add(TMP1, sp, 0); // Move that supports SP
-    stp(TMP1, lr, MemOperand(sp, -16, PreIndex));
+    // If we have a gdb server running then run in a less efficient mode that checks if we need to exit
+    // This happens when single stepping
+
+    static_assert(sizeof(CTX->Config.RunningMode) == 4, "This is expected to be size of 4");
+    ldr(x0, MemOperand(STATE, offsetof(FEXCore::Core::InternalThreadState, CTX)));
+    ldr(w0, MemOperand(x0, offsetof(FEXCore::Context::Context, Config.RunningMode)));
+    
+    // If the value == 0 then we don't need to stop
+    cbz(w0, &RunBlock);
+    {
+      // Make sure RIP is syncronized to the context
+      LoadConstant(x0, HeaderOp->Entry);
+      str(x0, MemOperand(STATE, offsetof(FEXCore::Core::ThreadState, State.rip)));
+
+      // Stop the thread
+      LoadConstant(x0, ThreadPauseHandlerAddressSpillSRA);
+      br(x0);
+    }
+    bind(&RunBlock);
   }
 
-  PendingTargetLabel = nullptr;
+  if (HeaderOp->ShouldInterpret) {
+    LoadConstant(x0, ThreadSharedData.InterpreterFallbackHelperAddress);
```

Comment:
Why does this change to indirect rather than just passing back `ThreadSharedData.InterpreterFallbackHelperAddress`?
This adds a small trampoline to every interpreter fallback for no reason that I can see.

Reply:
Because of gdb stub header, and later on of unsynchronized (implicit) RIP